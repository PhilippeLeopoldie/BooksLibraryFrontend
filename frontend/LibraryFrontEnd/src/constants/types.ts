export type BookType = {
    id: number;
    title: string;
    author: string;
    description: string;
    imageUrl?: string;
    averageRate: number;
    opinions?: ReviewType[] | null;
    genreId?: string
};

export type GenreType = {
    id: string,
    name: string,
    isForStoryGeneration?: boolean,
    books?: BookType[]
};

export type PaginatedBookType = {
    paginatedItems: BookType[],
    totalItems: number,
    page: number,
    totalPages: number,
    requestedAt: string
}

export type ReviewType = {
    id: number;
    rate: number;
    view: string;
    userName: string;
    postDate: string;
    bookId: number;
};

export type PaginationType = {
    page: string,
    pageSize: string
}
//Dynamically infer AistoryLanguageType from languages values
export type AiStoryLanguageType = typeof languages[number];

 export const languages = ["English", "French", "Swedish", "Spanish", "Italian", "Turkish"] as const;
