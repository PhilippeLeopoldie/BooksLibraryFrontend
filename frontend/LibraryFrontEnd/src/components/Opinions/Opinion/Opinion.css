.opinion__loading {
  background-color: transparent; 
  margin: 10%;
}
.opinion__loading--black {
  color: #ffffff;
}

.opinionContainer {
  display: grid;
  border-bottom-right-radius: 5%;
  background-color: transparent;
  -webkit-logical-width:16rem;
  margin-left: 1rem;
}



.opinionContainer--antique {
  margin-left: 2.4rem;
}


.opinionCardItems {
  background-color: transparent;
}

.opinioncard__reviews--flex {
  display: flex;
  
}


.opinioncard__reviews__star {
  text-align: left;
  margin-top: -1rem;
}



.OpinionCard__LastReviewTitle {
  font-style: italic;
  text-align: left;
  font-size: 0.875rem;
  margin-bottom: 0.5rem;
}

.opinionCard__userName {
  text-align: left;
  font-weight: bold;
}

.opinioncard__rate--flex {
  display: flex;
}

.opinioncard--view {
  text-align: left;
  font-style: italic;
  width: 16rem;
  height: 4rem;
  border-color: transparent;
  resize: none;
  overflow: hidden;
  font-size: 1rem;
}

.opinioncard__postDate {
  margin-left: 0.5rem;
  text-align: left;
}



.opinioncard--footer {
  display: flex;
  align-items: end;
  justify-content: center;
  border-bottom-right-radius: 5%;
}

.opinioncard--buttonmodify {
  margin-bottom: 5%;
}
.opinioncard--buttonmodify:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}
