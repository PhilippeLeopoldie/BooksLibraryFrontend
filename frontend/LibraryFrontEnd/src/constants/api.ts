export const API_URL : string = "https://booklibrary-backend-20f7a19cecb2.herokuapp.com/api/" 
//export const API_URL : string = "https://localhost:7152/api/";
export const BOOK_URL: string = API_URL + "Books";
export const BOOK_TOP_BOOK_URL : string = BOOK_URL+"/TopBooks?numberOfBooks="
export const BOOK_BY_BOOKID_URL: string = BOOK_URL+"/";
export const BOOK_BY_TITLE_OR_AUTHOR_URL : string = BOOK_URL+"/TitleOrAuthor?titleOrAuthor=";
export const BOOKS_BY_GENRESId_URL: string = `${BOOK_URL}/genre?genresId=`
export const OPINION_URL: string = API_URL+"Opinions/";
export const OPINION_BY_BOOKID_URL : string = OPINION_URL+"BookId=";
export const GENRES_LIST_URL: string = API_URL + "Genres";
export const AI_STORY_URL: string = `${API_URL}Story/AI`;



