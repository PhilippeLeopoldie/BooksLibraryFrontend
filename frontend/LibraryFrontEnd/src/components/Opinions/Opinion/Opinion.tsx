import { BookType, ReviewType } from "../../../constants/types";
import { useEffect, useState } from "react";
import "./Opinion.css";
import { OPINION_BY_BOOKID_URL } from "../../../constants/api";


type BookAndReviewType = {
    book: BookType,
    displayReview: (opinions: ReviewType[]) => void;
};

export const Opinion = ({book,displayReview}: BookAndReviewType) => {
    
    const FetchOpinionBYBookId = async (bookId: number) => {
        try {
            const response: Response = await fetch(
                OPINION_BY_BOOKID_URL + bookId
            );
            if (response.status === 200) {
                const responseData = await response.json();
                displayReview(responseData);
            } else if (response.status === 404) {
                console.log("404 response", response);
            }
        } catch (error) {
            console.error("error fetching Opinions:", error);
        }
    };

    useEffect(() => {
        FetchOpinionBYBookId(book.id);
    }, []);

    return (
        <>
        </>
    );
};
