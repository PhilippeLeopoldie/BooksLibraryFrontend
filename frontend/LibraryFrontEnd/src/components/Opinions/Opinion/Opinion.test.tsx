import { BookType } from "../../../constants/types";
import { Opinion } from "./Opinion";
import { act, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
//import { Book } from "../../Books/Book/Book";

global.fetch = jest.fn().mockImplementation(() =>
  Promise.resolve({
    json: () =>
      Promise.resolve([
        { Opinion: { id: 1, view: "View 1", userName: "userName 1", rate: 2 } },
      ]),
    status: 200,
    hearders: new Headers(),
  })
);

const book :BookType = {
  id:1,
  title:"title 1",
    author: "author 1",
    description: "description 1",
    averageRate:5
}

describe("Opinions Component Tests", () => {
  it("SnapShot: should render Opinion", () => {
    const {asFragment} =  
      render(<Opinion 
        book={book}       
        displayReview={()=>{}}/>);
    expect(asFragment()).toMatchSnapshot();
  })

  it("Should not render Loading message when book with no opinion", async () => {
    //Arrange
    
    await act(async () => {
      render(<Opinion 
        book={book} 
        displayReview={()=>{}} />);
    });
  
    //Act
    const loadingMessage = screen.queryByText("Loading...");
  
    //Expect
    expect(loadingMessage).toBeNull();
  }); 
})

