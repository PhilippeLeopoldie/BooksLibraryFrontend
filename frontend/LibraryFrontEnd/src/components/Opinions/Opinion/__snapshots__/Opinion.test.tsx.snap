// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Opinions Component Tests SnapShot: should render Opinion 1`] = `<DocumentFragment />`;
